File: Algorithms/p220_Contains_Duplicate_III/p220_Contains_Duplicate_III.cpp
class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
        if(nums.empty()) return 0;
        int pos = min(int(nums.size()),k+1);
        set<int> s{nums[0]};
        for(int i=1;i<nums.size();++i){
            // erase before insert for cases where nums[i-k-1]==nums[i]
            if(i>=pos) s.erase(nums[i-k-1]);
            auto res = s.insert(nums[i]);
            if(!res.second&&t>=0){
                return 1;
            }else{
                if(res.first!=s.begin()&&static_cast<long long int>(*res.first)-static_cast<long long int>(*prev(res.first))<=static_cast<long long int>(t)) return 1;
                if(next(res.first)!=s.end()&&static_cast<long long int>(*next(res.first))-static_cast<long long int>(*res.first)<=static_cast<long long int>(t)) return 1;
            }
        }
        return 0;
    }
};

File: Algorithms/p220_Contains_Duplicate_III/p220_Contains_Duplicate_III.py
class Solution(object):
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        """
        :type nums: List[int]
        :type k: int
        :type t: int
        :rtype: bool
        """
        if t<0:
            return False
        d = dict()
        w = t+1
        for i in range(len(nums)):
            if i>k:
                key = nums[i-k-1]//w
                del d[key]
            key = nums[i]//w
            if key in d:
                return True
            if key-1 in d and nums[i]-d[key-1]<=t:
                return True
            if key+1 in d and d[key+1]-nums[i]<=t:
                return True
            d[key] = nums[i]
        return False
